### PR TITLE
internal:Happy eyeballs

### PR DESCRIPTION
**Pull request by larry-safran**
Fix the Happy Eyeballs work (pr #10486) done by Tony on PickFirstLeafLoadBalancer and enhance tests.

### CODE REVIEW

**Comment by larry-safran** [comment]
Rebased onto master.  @YifeiZhuang it is ready to review.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 1]
```diff
@@ -225,15 +253,52 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
...
+        } else if (!addressIndex.isValid()) {
+          if (concludedState == TRANSIENT_FAILURE) {
+            // sticky TRANSIENT_FAILURE
+            helper.refreshNameResolution();
+            updateBalancingState(TRANSIENT_FAILURE,
+                new Picker(PickResult.withError(stateInfo.getStatus())));
+            break;
+          }
+
+          boolean hasIdle = false;
```
Comment (YifeiZhuang) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:278]: I might miss something, could you point me to the section in grfc about the hasIdle/hasConnecting thing.
Also what is missing here is to count the TF times starting from the first pass based on grfc `Happy Eyeballs in the pick_first LB Policy
` section

> If the first pass completes without a successful connection attempt, ... We will count the number of connection failures, and when that number reaches the number of subchannels, we will request re-resolution;

**Review by YifeiZhuang — COMMENTED** [review]
Sending what I have. One question is that it looks not easy to make the happy eyeball logics controlled by a flag.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 2]
```diff
@@ -324,6 +432,9 @@ private Subchannel createNewSubchannel(SocketAddress addr) {
             new EquivalentAddressGroup(addr)))
         .addOption(HEALTH_CONSUMER_LISTENER_ARG_KEY, hcListener)
             .build());
+    if (subchannel == null) {
+      throw new IllegalArgumentException("Was not able to create subchannel for " + addr);
```
Comment (YifeiZhuang) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:436]: IIRC LB should never throw. Otherwise it cause channel panic.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 3]
```diff
@@ -85,62 +93,68 @@ public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
...
         Collections.shuffle(servers,
             config.randomSeed != null ? new Random(config.randomSeed) : new Random());
       }
     }
 
+    // Make sure we're storing our own list rather than what was passed in
     final List<EquivalentAddressGroup> newImmutableAddressGroups =
         Collections.unmodifiableList(new ArrayList<>(servers));
 
+    boolean reuseReadyChannel = false;
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:114]: done

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 4]
```diff
@@ -85,62 +93,68 @@ public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
...
+    // Shut them down and remove them
     for (SocketAddress oldAddr : oldAddrs) {
       if (!newAddrs.contains(oldAddr)) {
-        subchannels.get(oldAddr).getSubchannel().shutdown();
-        subchannels.remove(oldAddr);
+        subchannels.remove(oldAddr).getSubchannel().shutdown();
       }
     }
 
+    if (reuseReadyChannel) {
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:148]: done

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 5]
```diff
@@ -179,6 +194,19 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
     if (newState == SHUTDOWN) {
       return;
     }
+
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:197]: I don't follow how the subchannelData check is covering the passed in state value not being SHUTDOWN.  Can you clarify?

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 1 above)
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:285]: Updated to only do the refreshNameResolution when there have been sufficient connection failures.

Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:285]: Happy Eyeballs should be the same as the petiole policies which had the rule that if any child is READY, then it is READY, if not but any child is CONNECTING then it is CONNECTING, if not but some child is IDLE then it is IDLE, otherwise it is TRANSIENT FAILURE.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 6]
```diff
@@ -291,29 +357,71 @@ private void shutdownRemaining(SubchannelData activeSubchannelData) {
...
+    ConnectivityState subchannelState = subchannels.get(currentAddress).getState();
+    switch (subchannelState) {
+      case IDLE:
         subchannel.requestConnection();
-      } else if (subchannelState == CONNECTING || subchannelState == TRANSIENT_FAILURE) {
+        subchannels.get(currentAddress).updateState(CONNECTING);
+        scheduleNextConnection();
+        break;
+      case CONNECTING:
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:398]: This is in `requestConnection`, not in `processSubchannelState`.   We have already made the decision that we are requesting a connection and are just trying to find the appropriate subchannel to request it on.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 7]
```diff
@@ -291,29 +357,71 @@ private void shutdownRemaining(SubchannelData activeSubchannelData) {
...
+      case TRANSIENT_FAILURE:
         addressIndex.increment();
         requestConnection();
+        break;
+      case READY: // Since we aren't doing anything there won't be a state change
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java]: done

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 2 above)
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java]: I can return a null, and then in requestConnection try the next address.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 1 above)
Comment (YifeiZhuang) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:286]: Happy eyeballs is pick first not petiole policies. I don't think aggregated connectivity applies here this way. The aggregated connectivity processing here makes it unnecessarily complicated.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (YifeiZhuang) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:399]: `requestConnection` can be from both. Upon address update, or as a result of `processSubchannelState`.
grfc made it clear that when we requesting a connection, this subchannel may NOT be in IDLE. 
> If the subchannel is in state CONNECTING, we do not need to actually request a connection, but we will treat it as if we did. If it is not the last subchannel in the list, we will start the timer; if it is the last subchannel in the list, we will wait for the attempt to complete.

So here we should not increment the index, instead we should restart the timer if it is not the last address.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (YifeiZhuang) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:198]: we already talked offline that when rawConnectivityState is READY, there is only one active subchannel remained, others are all shutdown. And L191 ignores callback from other subchannels different from the current one.

**Comment by larry-safran** [comment]
@YifeiZhuang Addressed all of the comments including incorporating a flag to enable the logic.  Mostly, disabling is just not scheduling the try of the next subchannel on timeout.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 8]
```diff
@@ -225,15 +244,40 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
...
+          if (!addressIndex.isValid()
+              && (!enableHappyEyeballs || numTfSinceAccept >= addressIndex.size())) {
+            // If no addresses remaining, go into TRANSIENT_FAILURE
             helper.refreshNameResolution();
             rawConnectivityState = TRANSIENT_FAILURE;
             updateBalancingState(TRANSIENT_FAILURE,
                 new Picker(PickResult.withError(stateInfo.getStatus())));
+            break;
+          }
+        } else if (!addressIndex.isValid()) {
```
Comment (YifeiZhuang) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:261]: I feel this else-if block is hard to follow and needs cleanup. Isn't this enough:
```
else if (!addressIndex.isValid() && enableHappyEyeball && ++counter) {
  refresh name resolution.
} 
```

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 9]
```diff
@@ -291,29 +343,92 @@ private void shutdownRemaining(SubchannelData activeSubchannelData) {
...
-          subchannels.get(addressIndex.getCurrentAddress()).getState();
-      if (subchannelState == IDLE) {
+    Subchannel subchannel;
+    SocketAddress currentAddress;
+    do {
+      currentAddress = addressIndex.getCurrentAddress();
+      subchannel = subchannels.containsKey(currentAddress)
+          ? subchannels.get(currentAddress).getSubchannel()
+          : createNewSubchannel(currentAddress);
+    } while (subchannel == null && addressIndex.isValid());
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java]: Done.  Updated helper.createSubchannel to mention that null is not allowed and createNewSubchannel to throw an exception instead of returning null.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 10]
```diff
@@ -291,29 +343,92 @@ private void shutdownRemaining(SubchannelData activeSubchannelData) {
...
+    switch (subchannelState) {
+      case IDLE:
         subchannel.requestConnection();
-      } else if (subchannelState == CONNECTING || subchannelState == TRANSIENT_FAILURE) {
+        subchannels.get(currentAddress).updateState(CONNECTING);
+        scheduleNextConnection();
+        break;
+      case CONNECTING:
+        if (enableHappyEyeballs) {
+          if (!addressIndex.isAtEnd()) {
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java]: Done

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 11]
```diff
@@ -324,19 +439,18 @@ private Subchannel createNewSubchannel(SocketAddress addr) {
             new EquivalentAddressGroup(addr)))
         .addOption(HEALTH_CONSUMER_LISTENER_ARG_KEY, hcListener)
             .build());
+    if (subchannel == null) {
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:417]: If the helper is created correctly.  I would prefer not to trust something that we don't completely control and do the check.  However, I am now throwing an exception rather than returning null.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 12]
```diff
@@ -55,11 +57,20 @@
...
 final class PickFirstLeafLoadBalancer extends LoadBalancer {
   private static final Logger log = Logger.getLogger(PickFirstLeafLoadBalancer.class.getName());
+  @VisibleForTesting
+  static final int CONNECTION_DELAY_INTERVAL_MS = 250;
+  public static final String GRPC_EXPERIMENTAL_XDS_DUALSTACK_ENDPOINTS =
+      "GRPC_EXPERIMENTAL_XDS_DUALSTACK_ENDPOINTS";
   private final Helper helper;
   private final Map<SocketAddress, SubchannelData> subchannels = new HashMap<>();
   private Index addressIndex;
+  private int numTfSinceAccept = 0;
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java]: Changed the logic to only start counting after the first pass is complete and added a test to make sure that we are following the grfc.

[the remark this answers is not in the document]
Comment (YifeiZhuang) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java]: I would correct a wrong statement I mentioned above. Both acceptResolvedAddress and processSubchannelState (see [callback](https://github.com/grpc/grpc-java/blob/f01c110f7788d79ba87e5f37c82a320f8c35f1a3/core/src/main/java/io/grpc/internal/InternalSubchannel.java#L73) in InternalSubchannel) are in the same syncContext in managedChannelImpl, although they might from different threads. So there is no thread safety issue. And the `AddressIndex` can just be local without synchronized.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 13]
```diff
@@ -211,32 +231,45 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
...
+          if (addressIndex.increment()) {
+            cancelScheduleTask();
+            requestConnection(); // is recursive so might hit the end of the addresses
           }
         }
+
+        // If at end and: not doing happy eyeballs, just got to the end, or we have had enough TF
+        // since we got to the end, then refresh name resolution
+        if (!addressIndex.isValid()
+            && (!enableHappyEyeballs || numTf.get() >= addressIndex.size() || numTf.get() == 0)) {
```
Comment (YifeiZhuang) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:260]: I see that you combined a few cases here. It's not trivial to verify and hard to maintain. 
For example, it looks wrong in the case of `enableHappyEyeballs==false`, because name resolution is refreshed every time when addressIndex is invalid and we have TF. 
Also I think increment the counter and the action should be together instead of scatter around. What about move L198-L200 down, and what about call `helper.refreshNameresolution` twice, once for the first time it goes invalid, then every time the counter > address size.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java]: Done

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 14]
```diff
@@ -85,54 +100,57 @@ public Status acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
         return unavailableStatus;
       }
     }
+
+    // Since we have a new set of addresses, we are again at first pass
+    numTf.getAndSet(0);
```
Comment (YifeiZhuang) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:105]: I think remove these two lines. Because it is much simpler to count the numTF after first pass, instead of start from address update. That is easier to reason about. We have all the conditions to tell it is first pass so we do not need extra variable. See the code snippet in the later comments.
When we have address update, we reset the index already, (only case is READY connection and address still in the new list, when the READY connection goes IDLE, it will reset the index, then it comes to the same situation), so when the first time index goes to invalid when we iterating the list, we can tell it is first pass complete. 
In case that when we receive the exact same set of addresses and they are all in TF already, it will miss the first pass but it is fine because the it is name resolution is already refreshed before update.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 15]
```diff
@@ -211,32 +230,49 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
...
         updateHealthCheckedState(subchannelData);
         break;
+
       case TRANSIENT_FAILURE:
         // If we are looking at current channel, request a connection if possible
         if (addressIndex.isValid()
             && subchannels.get(addressIndex.getCurrentAddress()).getSubchannel() == subchannel) {
-          addressIndex.increment();
-          requestConnection();
+          if (addressIndex.increment()) {
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:250]: There are 2 other cases where we hit the end  of addressIndex and for both of them we don't know that a connection attempt has completed:

- the happy eyeballs scheduled task's deadline being exceeded (`StartNextConnection.run`)
- `RequestConnectionPicker.pickSubchannel`

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 16]
```diff
@@ -211,32 +230,49 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
...
+          }
+        }
 
-          // If no addresses remaining, go into TRANSIENT_FAILURE
-          if (!addressIndex.isValid()) {
+        // If we are (at the end and have had enough TF since we got there) or
+        // (just got to the end), then refresh name resolution which may call
+        // acceptResolvedAddresses which will reset addressIndex, but may not.
+        if (!addressIndex.isValid()) {
+          numTf.getAndIncrement();
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:260]: It looks like `numTF` doesn't need to be atomic. And that's good, because these three calls to adjust it wouldn't be thread-safe. We should make it a normal integer.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 17]
```diff
@@ -55,11 +58,21 @@
...
   private final Helper helper;
   private final Map<SocketAddress, SubchannelData> subchannels = new HashMap<>();
   private Index addressIndex;
+  private AtomicInteger numTf = new AtomicInteger(0);
+  private volatile boolean firstPass = true;
+  @Nullable
+  private ScheduledHandle scheduleConnectionTask;
   private ConnectivityState rawConnectivityState = IDLE;
   private ConnectivityState concludedState = IDLE;
+  private final boolean enableHappyEyeballs =
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:74]: static?

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (YifeiZhuang) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:250]: 1. You are right. I see that the "first pass complete" check I proposed is not sufficient in the case where they are incremented somewhere else.
One concern I have is that in those cases where the first pass is done happened in other places instead of `processSubchannelState`, we are checking `firstPass` at the next subchannel failure, which is too late, worse case is that the transport comes back until it times out.  I think the way C++ is doing is more accurate, https://github.com/grpc/grpc/blob/6a4b5ccea362dd47562c698dd2aa5a8aead41a6a/src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.cc#L1150 what do you think?
2. This is minor but you could have let the numTF only increment when first pass is done. But either way.
3. Add `enableHappyEyeballs` condition.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 16 above)
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:260]: Changed to a normal int.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 17 above)
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:74]: I wasn't able to think of a clean way to set it in tests if it was static.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:250]: 1. We need to check when the connection attempt completes which is when it calls processSubchannelState, so I believe that is the appropriate way to do it.  I've updated the logic to actually look at the subchannel states to determine that they have all completed.
2. I believe that the new logic using passComplete() does this
3. The old PickFirstLoadBalancer calls refreshNameResolution on every TF it receives [code](https://github.com/grpc/grpc-java/blob/be74bbdfe0d8a799e7fc32155bb0772371ab8b4c/core/src/main/java/io/grpc/internal/PickFirstLoadBalancer.java#L116-L118).  I agree with an earlier comment you made that name resolution should be refreshed like in the happy eyeballs case only at end of first pass and after address count worth of TF status reports, though @ejona86 may disagree.

[the remark this answers is not in the document]
Comment (YifeiZhuang) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:248]: 1.   When `requestConnection` from `pick()` or immediate after `acceptResolvedAddress` reaches the last address, it won't call back `processSubchannelState`,  e.g. when all subchannel in TF status already.  I believe the behaviour won't be a huge disaster but still different.
3. The old PickFirstLoadBalancer only has one subchannel but the iteration logic is inside one subchannel, so it calls refreshNameResolution every TF actually means every iteration of all addresses. It looks before your happy eyeballs change it only refreshNameResolution once, which seems a bug. I'm ok with your change this way unless Eric has different views.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java [hunk 18]
```diff
@@ -211,32 +228,48 @@ void processSubchannelState(Subchannel subchannel, ConnectivityStateInfo stateIn
...
+              new Picker(PickResult.withError(stateInfo.getStatus())));
+
+          // Refresh Name Resolution, but only when all 3 conditions are met
+          // * We are at the end of addressIndex
+          // * have had status reported for all subchannels.
+          // * And one of the following conditions:
+          //    * Have had enough TF reported since we completed first pass
+          //    * Just completed the first pass
+          if (++numTf >= addressIndex.size() || firstPass) {
+            firstPass = false; // idempotent so okay to call when false
```
Comment (YifeiZhuang) [inline core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:266]: Nit: I know what you meant by saying `idempotent` but technically the word is a more strict terminology. Just remove the comment would be clear.

File: core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java [hunk 19]
```diff
@@ -110,46 +125,53 @@ public void uncaughtException(Thread t, Throwable e) {
   private FakeSubchannel mockSubchannel3;
   @Mock
   private FakeSubchannel mockSubchannel4;
+  @Mock
+  private FakeSubchannel mockSubchannel5;
   @Mock // This LoadBalancer doesn't use any of the arg fields, as verified in tearDown().
   private PickSubchannelArgs mockArgs;
 
+  @BeforeClass
+  public static void setUpClass() {
```
Comment (YifeiZhuang) [inline core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java:134]: I didn't see you reset the property @AfterClass, so other tests in the entire grpc test will always see happyeyeballs enabled. However I think we expect them to test on happy eyeballs not enabled case.
More importantly, the test suite here should verify both enabled/disabled cases (use Parameterized), or at least have the original tests do the disabled cases, and have your new tests enable the variable. I didn't realize last time I saw.

File: core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (YifeiZhuang) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:252]: I think this still remains an issue though:

> When requestConnection from pick() or immediately after acceptResolvedAddress reaches the last address in its iteration, `processSubchannelState` does not come back, e.g. when all subchannel in TF status already. This would cause the refreshNameResolution delayed.

If you have different opinion, let's engage Eric in the thread. Otherwise at least add comment to fixit.

[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java:253]: I created a couple of tests `happy_eyeballs_connection_results_happen_after_get_to_end` and `happy_eyeballs_pick_pushes_index_over_end` which try these use cases and they don't have any issues.

File: core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java [hunk 20]
```diff
@@ -110,46 +121,55 @@ public void uncaughtException(Thread t, Throwable e) {
...
   private FakeSubchannel mockSubchannel4;
+  @Mock
+  private FakeSubchannel mockSubchannel5;
   @Mock // This LoadBalancer doesn't use any of the arg fields, as verified in tearDown().
   private PickSubchannelArgs mockArgs;
 
   @Before
   public void setUp() {
-    for (int i = 1; i < 5; i++) {
+    System.setProperty(PickFirstLeafLoadBalancer.GRPC_EXPERIMENTAL_XDS_DUALSTACK_ENDPOINTS, "TRUE");
```
[the remark this answers is not in the document]
Comment (larry-safran) [reply core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java]: Done.

File: core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java (hunk 19 above)
Comment (larry-safran) [reply core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java]: Changed to setting in `setup` and added an `@after` method to restore it.

File: core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java [hunk 21]
```diff
@@ -2057,15 +2076,264 @@ public void ready_then_transient_failure_again() {
...
+    pickResult = requestingPicker.pickSubchannel(mockArgs);
+    assertEquals(PickResult.withNoResult(), pickResult);
+
+    // Sending TF state to one subchannel pushes index past end, but shouldn't do anything
+    listeners[2].onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
+    inOrder.verify(mockHelper, never()).updateBalancingState(eq(TRANSIENT_FAILURE), any());
+
+    // Put the LB into TF
+    listeners[0].onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
+    listeners[1].onSubchannelState(ConnectivityStateInfo.forTransientFailure(error));
```
Comment (YifeiZhuang) [inline core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java:2225]: One example is here, say we have 3 subchannels(TF, TF, TF). Index points to invalid.
Then we have address update with exactly the same addresses. So we reset FirstPass and Index = 0. 
We start do requestConnection() which iterates all 3 addresses. 
After that we should have a `refreshNameResolution` but current behaviour won't, because that logic in only in the listener, which is not sufficient. Right? I understand this may complicate the logic, so if we do not have a simple solution we should comment.

Comment (larry-safran) [reply core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java:2225]: Since the subchannels are in TF, each of them is retrying their connections independently of the load balancer.  We reset firstPass to true, so as soon as one of those retries gets a TF response we'll call refreshNameResolution.

Comment (YifeiZhuang) [reply core/src/test/java/io/grpc/internal/PickFirstLeafLoadBalancerTest.java:2225]: Talked offline about the (potentially delayed, but purposely) `refreshNameResolution` behaviour when all subchannels are in TF upon address update, or part of the subchannels are in TF upon address update. This looks fine to me.

**Review by YifeiZhuang — APPROVED** [review]
LGTM
